browser/brave_wallet/brave_wallet_provider_delegate_impl.cc

```diff
@@ -19,4 +51,45 @@ void BraveWalletProviderDelegateImpl::ShowConnectToSiteUI() {
   brave::ShowWalletBubble(browser);
 }
 
+void BraveWalletProviderDelegateImpl::RequestEthereumPermissions(
+    RequestEthereumPermissionsCallback callback) {
+  permissions::PermissionManager* permission_manager =
+      permissions::PermissionsClient::Get()->GetPermissionManager(
+          web_contents_->GetBrowserContext());
+  if (!permission_manager) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  auto* service = BraveWalletServiceFactory::GetInstance()->GetForContext(
+      web_contents_->GetBrowserContext());
+  auto* keyring_controller = service->keyring_controller();
+  auto* default_keyring = keyring_controller->GetDefaultKeyring();
+  std::vector<std::string> addresses;
+  if (default_keyring) {
+    addresses = default_keyring->GetAccounts();
+  }
+
+  // To support ethereum permission to be per Ethereum account per site, we map
+  // each account address to one ethereum permission request. And the requests
+  // will have different origins which includes the address information. Here
+  // we first get a concatenated origin to include information of all wallet
+  // addresses, then adjust the origin of each request later in the process
+  // because PermissionManager::RequestPermissions only accepts a single origin
+  // parameter to be passes in.
+  GURL origin;
+  if (!GetConcatOriginFromWalletAddresses(
+          web_contents_->GetLastCommittedURL().GetOrigin(), addresses,
+          &origin)) {
+    // Failed due to origin is invalid or addresses is empty.
+    std::move(callback).Run(false);
+    return;
+  }
+  std::vector<ContentSettingsType> types(addresses.size(),
+                                         ContentSettingsType::BRAVE_ETHEREUM);
+  permission_manager->RequestPermissions(
+      types, web_contents_->GetMainFrame(), origin, true,
```

Comment:
we need to make sure we are passing the actual user gesture status here so sites can't abuse this

Reply:
I think `web_contents_->HasRecentInteractiveInputEvent` should suffice

Reply:
actually we need the rfh here anyway to make sure we have the correct url for the permissions prompt so `rfh->HasTransientUserActivation`

Reply:
Fixed in fc9e5ed